src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -4397,18 +4324,21 @@ def _contains_intrinsic_reduction(pnodes):
         # parent for this logical expression we will repeat the processing.
         fake_parent = Assignment(parent=parent)
         self.process_nodes(fake_parent, logical_expr)
+        # We want to convert all the non-explicit array syntax code to use
+        # explicit array syntax.
+        # TODO 1799 If we have two arrays where one uses a non-maximal range
+        # that is defined we will not convert this correctly yet.
+        # Convert References to arrays to use the array range notation unless
+        # they have an IntrinsicCall parent.
+        for ref in fake_parent.walk(Reference):
+            if (isinstance(ref.symbol, DataSymbol) and
+                    not ref.ancestor(IntrinsicCall)):
```

Comment:
Why it the IntrinsicCall check needed? If this is just related to "inquiry intrinsics" see next comment. But regardless, I would like to see a test with an inquiry, an elemental and a non-elemental intrinsic inside a where to make sure we do the right thing for each case. (if it's hard, placing them in codeblocks is good enough)

Reply:
I saw something saying we don't allowed intrinsics (other than one specific one) inside `WHERE` statements, presumably because if you had something like `WHERE(a(:) > b(:)) c = MAX(a(:))` (not sure of exact syntax) then we'd need to do quite a lot more work to support this, so currently we disallow all but one specific one which already has a test i think. The others all just should result in a codeblock I think, but I'll try to put together a test.

Reply:
I'm not sure which intrinsics make sense from a Fortran perspective inside a WHERE clause, I'll try a few things out and see if it works.

Reply:
maxval doesn't work for example.

Reply:
> I saw something saying we don't allowed intrinsics (other than one specific one)

If intrinsics are not supported, that would be fine, but here we skip the transformation if we fine a intrinsics, which is what I don't understand. I think the key is if they are elemental or not. For instance in this example:
```
    where(a > 1)
        b = abs(c) + sum(c)
    end where
```
we currently convert it to:
```
  do widx1 = 1, 10, 1
    if (a(widx1) > 1) then
      b(widx1) = ABS(c) + SUM(c)
    end if
  enddo
```
Which is wrong, we generate code that won't compile! It should be `b(widx1) = ABS(c(widx1)) + SUM(c)`

Reply:
Technically, the issue is with function calls (which we need to know if they are elemental or not) not with intrinsics. If its not easy to solve we could fail if the expression has any call (and let it be a codeblock).

Reply:
Ah ok I see, if I find an intrinsic that isn't `SUM` then I'll raise a NotImplementedError for now.

Reply:
I'm actually not sure - I think intrinsics are fine except for the ones blocked by `_contains_intrinsic_reduction` so I'm going to remove this.

Reply:
Added a test using SIN - should now work as expected.